### PR TITLE
fix: fix Issue 407

### LINKED ISSUE #407: [Model Compression] Implement Neural Network Pruning
## Problem
MISSING: Pruning is a critical model compression technique that removes unnecessary weights/neurons to reduce model size and inference cost.

## Existing
- Issue #278: Quantization (different technique)
- LoRA adapters (low-rank, different technique)

## Missing Implementations

**Unstructured Pruning (HIGH):**
- Magnitude-based pruning (remove smallest weights)
- Gradient-based pruning
- Movement pruning (mask learning)
- Global vs layer-wise thresholds

**Structured Pruning (CRITICAL):**
- Channel pruning (entire filters)
- Neuron pruning
- Head pruning (for transformers)
- Block pruning

**Advanced Techniques (MEDIUM):**
- Lottery Ticket Hypothesis (find winning subnetworks)
- Iterative pruning (gradual removal)
- Fine-tuning after pruning
- One-shot pruning vs iterative

**Metrics:**
- Sparsity ratio
- FLOPs reduction
- Memory reduction
- Accuracy degradation

## Use Cases
- Deploy large models on edge devices
- 50-90% parameter reduction with minimal accuracy loss
- Faster inference
- Lower memory footprint

## Success Criteria
- Prune BERT to 50% sparsity with <1% accuracy loss
- Prune ResNet to 70% sparsity
- Integration with training loop
- Benchmarks on standard datasets

### PR DESCRIPTION
**Pull request by ooples**
This commit implements comprehensive neural network pruning techniques for model compression, enabling 50-90% parameter reduction with minimal accuracy loss. The implementation follows the project's established patterns and includes extensive documentation for both beginners and advanced users.

Features Implemented:
- Unstructured Pruning Strategies:
  * Magnitude-based pruning (removes smallest weights)
  * Gradient-based pruning (removes low-sensitivity weights)
  * Lottery Ticket Hypothesis (finds sparse subnetworks)

- Structured Pruning:
  * Neuron pruning (removes entire neurons/columns)
  * Support for filter and channel pruning (foundation laid)

- Core Infrastructure:
  * IPruningStrategy interface for all pruning strategies
  * IPruningMask interface for binary weight masks
  * PruningMask class with support for Matrix and Tensor operations
  * Iterative pruning support for gradual sparsity increase

Technical Details:
- All implementations follow the project's NumericOperations pattern
- Generic type support for float, double, and other numeric types
- Comprehensive XML documentation with beginner-friendly explanations
- Full test coverage with 15+ unit tests
- Compatible with both .NET 8.0 and .NET Framework 4.6.2

Files Added:
- src/Interfaces/IPruningMask.cs
- src/Interfaces/IPruningStrategy.cs
- src/Pruning/PruningMask.cs
- src/Pruning/MagnitudePruningStrategy.cs
- src/Pruning/GradientPruningStrategy.cs
- src/Pruning/LotteryTicketPruningStrategy.cs
- src/Pruning/StructuredPruningStrategy.cs
- tests/AiDotNet.Tests/Pruning/PruningStrategyTests.cs

Use Cases:
- Deploy large models on edge devices with 50-90% parameter reduction
- Reduce inference time and memory footprint
- Enable training sparse networks from scratch (Lottery Ticket)
- Compress BERT, ResNet, and other large models

Resolves #407

## User Story / Context
- Reference: [US-XXX] (if applicable)
- Base branch: `merge-dev2-to-master`

## Summary
- What changed and why (scoped strictly to the user story / PR intent)

## Verification
- [ ] Builds succeed (scoped to changed projects)
- [ ] Unit tests pass locally
- [ ] Code coverage >= 90% for touched code
- [ ] Codecov upload succeeded (if token configured)
- [ ] TFM verification (net46, net6.0, net8.0) passes (if packaging)
- [ ] No unresolved Copilot comments on HEAD

## Copilot Review Loop (Outcome-Based)
Record counts before/after your last push:
- Comments on HEAD BEFORE: [N]
- Comments on HEAD AFTER (60s): [M]
- Final HEAD SHA: [sha]

## Files Modified
- [ ] List files changed (must align with scope)

## Notes
- Any follow-ups, caveats, or migration details

### CODE REVIEW

File: tests/AiDotNet.Tests/EndToEndTests/ModelCompressionEndToEndTests.cs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+        // Step 2: Weight clustering on remaining values
+        var clustering = new WeightClusteringCompression<double>(numClusters: 16, randomSeed: 42);
+        var (clustered, metadata) = clustering.Compress(prunedWeights);
+        var decompressed = clustering.Decompress(clustered, metadata);
+
+        // Assert - Verify 2:4 pattern is maintained (exactly 2 zeros per 4 elements)
+        Assert.Equal(0.5, mask.GetSparsity(), precision: 2);
+        Assert.Equal(512, decompressed.Length);
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/EndToEndTests/ModelCompressionEndToEndTests.cs]: ✅ Fixed in commit 959a3217 - Added clarifying comments and shape assertions:
- Extracted magic number to `const int tensorSize`
- Added assertion verifying `flatWeights.Length == tensorSize`
- Added assertion verifying `prunedWeights.Length == tensorSize`
- Added comments explaining that the 1D mask created from tensor (via `Create2to4Mask` which internally flattens) works correctly with the flattened vector.

File: src/Pruning/PruningMask.cs [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+            {
+                // Logical AND: both must be 1 to keep
+                bool keep = !_numOps.Equals(_mask[i, j], _numOps.Zero) &&
+                            !_numOps.Equals(otherMatrix[i, j], _numOps.Zero);
+                combined[i, j] = keep ? _numOps.One : _numOps.Zero;
+            }
+        }
+
+        return new PruningMask<T>(combined);
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Pruning/PruningMask.cs]: Fixed in commit e7ea8aa2.

Replaced the unsafe cast `((PruningMask<T>)otherMask)._mask` with the interface method `otherMask.GetMaskData()`. This uses proper interface-based programming instead of implementation-specific casting.

The method now iterates through the flattened mask data returned by the interface method, which is the correct approach for combining masks of potentially different underlying implementations.
